sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsReflectorTest.java

```diff
@@ -0,0 +1,196 @@
+import static org.hamcrest.Matchers.not;
+
+import com.google.common.collect.ImmutableSet;
+import org.hamcrest.FeatureMatcher;
+import org.hamcrest.Matcher;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import java.util.Set;
+
+/**
+ * Unit tests for {@link PipelineOptionsReflector}.
+ */
+@RunWith(JUnit4.class)
+public class PipelineOptionsReflectorTest {
+  @Test
+  public void testGetOptionSpecs() throws NoSuchMethodException {
+    Set<PipelineOptionSpec> properties =
+        PipelineOptionsReflector.getOptionSpecs(SimpleOptions.class);
+
+    assertThat(properties, Matchers.hasItems(PipelineOptionSpec.of(
+        SimpleOptions.class, "foo", SimpleOptions.class.getDeclaredMethod("getFoo"))));
+  }
+
+  interface SimpleOptions extends PipelineOptions {
+    String getFoo();
+    void setFoo(String value);
+  }
+
+  @Test
+  public void testFiltersNonGetterMethods() {
+    Set<PipelineOptionSpec> properties =
+        PipelineOptionsReflector.getOptionSpecs(OnlyTwoValidGetters.class);
+
+    assertThat(properties, not(hasItem(hasName(isOneOf("misspelled", "hasParameter", "prefix")))));
+  }
+
+  interface OnlyTwoValidGetters extends PipelineOptions {
+    String getFoo();
+    void setFoo(String value);
+
+    boolean isBar();
+    void setBar(boolean value);
+
+    String gtMisspelled();
+    void setMisspelled(String value);
+
+    String getHasParameter(String value);
+    void setHasParameter(String value);
+
+    String noPrefix();
+    void setNoPrefix(String value);
+  }
+
+  @Test
+  public void testBaseClassOptions() {
+    Set<PipelineOptionSpec> props =
+        PipelineOptionsReflector.getOptionSpecs(ExtendsSimpleOptions.class);
+
+    assertThat(props, Matchers.hasItem(
+        allOf(hasName("foo"), hasClass(SimpleOptions.class))));
+    assertThat(props, Matchers.hasItem(
+        allOf(hasName("foo"), hasClass(ExtendsSimpleOptions.class))));
+    assertThat(props, Matchers.hasItem(
+        allOf(hasName("bar"), hasClass(ExtendsSimpleOptions.class))));
+  }
+
+  interface ExtendsSimpleOptions extends SimpleOptions {
+    @Override String getFoo();
+    @Override void setFoo(String value);
+
+    String getBar();
+    void setBar(String value);
+  }
+
+  @Test
+  public void testExcludesNonPipelineOptionsMethods() {
+    Set<PipelineOptionSpec> properties =
+        PipelineOptionsReflector.getOptionSpecs(ExtendsNonPipelineOptions.class);
+
+    assertThat(properties, not(hasItem(hasName("foo"))));
+  }
+
+  interface NoExtendsClause {
+    String getFoo();
+    void setFoo(String value);
+  }
+
+  interface ExtendsNonPipelineOptions extends NoExtendsClause, PipelineOptions {}
+
+  @Test
+  public void testExcludesHiddenInterfaces() {
+    Set<PipelineOptionSpec> properties =
+        PipelineOptionsReflector.getOptionSpecs(Hidden.class);
+
+    assertThat(properties, not(hasItem(hasName("foo"))));
+  }
+
+  @org.apache.beam.sdk.options.Hidden
```

Comment:
This was ported over from `PipelineOptionsFactoryTest`, which defines the below `Hidden` pipelines options interface. I've renamed it to `HiddenOptions` to avoid the name clash.